pkg/build/controller/build/build_controller.go

```diff
@@ -0,0 +1,850 @@
+	build := cur.(*buildapi.Build)
+	bc.enqueueBuild(build)
+}
+
+// enqueueBuild adds the given build to the queue.
+func (bc *BuildController) enqueueBuild(build *buildapi.Build) {
+	key, err := kcontroller.KeyFunc(build)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for build %#v: %v", build, err))
+		return
+	}
+	bc.queue.Add(key)
+}
+
+// enqueueBuildForPod adds the build corresponding to the given pod to the controller
+// queue. If a build is not found for the pod, then an error is logged.
+func (bc *BuildController) enqueueBuildForPod(pod *kapi.Pod) {
+	buildName := buildutil.GetBuildName(pod)
+	build, err := bc.buildStore.Builds(pod.Namespace).Get(buildName)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't retrieve build for pod %s/%s: %v", pod.Namespace, pod.Name, err))
+		return
+	}
+	bc.enqueueBuild(build)
+}
+
+// handleError is called by the main work loop to check the return of the handleBuild or handlePodDeleted
+// calls. If an error occurred, then the key is re-added to the queue unless it has been retried too many
+// times.
+func (bc *BuildController) handleError(err error, key interface{}) {
+	if err == nil {
+		bc.queue.Forget(key)
+		return
+	}
+
+	if strategy.IsFatal(err) {
+		glog.V(2).Infof("Will not retry fatal error for key %v: %v", key, err)
+		bc.queue.Forget(key)
+		return
+	}
+
+	if bc.queue.NumRequeues(key) < maxRetries {
+		glog.V(4).Infof("Retrying key %v: %v", key, err)
+		bc.queue.AddRateLimited(key)
+		return
+	}
+
+	glog.V(2).Infof("Giving up retrying %v: %v", key, err)
+	bc.queue.Forget(key)
+}
+
+// isBuildPod returns true if the given pod is a build pod
+func isBuildPod(pod *kapi.Pod) bool {
+	return len(buildutil.GetBuildName(pod)) > 0
+}
+
+// buildDesc is a utility to format the namespace/name and phase of a build
+// for errors and logging
+func buildDesc(build *buildapi.Build) string {
+	return fmt.Sprintf("%s/%s (%s)", build.Namespace, build.Name, build.Status.Phase)
+}
+
+// podDeletedUpdate returns a buildUpdate object to apply to a build that has had
+// its pod deleted
+func podDeletedUpdate() *buildUpdate {
+	update := &buildUpdate{}
+	update.setPhase(buildapi.BuildPhaseError)
+	update.setReason(buildapi.StatusReasonBuildPodDeleted)
+	update.setMessage(buildapi.StatusMessageBuildPodDeleted)
+	return update
+}
+
+// isValidTransition returns true if the given phase transition is valid
+func isValidTransition(from, to buildapi.BuildPhase) bool {
+	if from == to {
+		return true
+	}
+
+	switch {
+	case isTerminal(from):
+		return false
+	case from == buildapi.BuildPhasePending:
+		switch to {
+		case buildapi.BuildPhaseNew:
+			return false
+		}
+	case from == buildapi.BuildPhaseRunning:
+		switch to {
+		case buildapi.BuildPhaseNew,
+			buildapi.BuildPhasePending:
+			return false
+		}
+	}
+
+	return true
+}
+
+// isTerminal returns true if the given build phase is a terminal state
+func isTerminal(phase buildapi.BuildPhase) bool {
```

Comment:
I don't like that in some places "terminal" means (complete|failed|error|cancelled), and in some places it means the same minus (complete/failed && completiontimestamp == nil).  Seems like a recipe for logic errors.  Is it possible to audit and improve this at all now?